addons/website_sale/controllers/main.py

```diff
@@ -405,6 +405,140 @@ def old_product(self, product, category='', search='', **kwargs):
+            request.env['product.template'].browse(image_res_id).image_1920 = False
+        else:
+            request.env['product.image'].browse(image_res_id).unlink()
+
+    @http.route(['/shop/product/resequence-image'], type='json', auth='user', website=True)
+    def resequence_product_image(self, image_res_model, image_res_id, move):
+        # TODO: make test for this
+        if not (request.env.user.has_group('website.group_website_designer') or\
+            request.env.user.has_group('website.group_website_publisher')) or\
+            image_res_model not in ['product.product', 'product.template', 'product.image'] or\
+            move not in ['first', 'left', 'right', 'last']:
+            raise NotFound()
+
+        image_res_id = int(image_res_id)
+        image_to_resequence = request.env[image_res_model].browse(image_res_id)
+        product = request.env['product.product']
+        product_template = request.env['product.template']
+        if image_res_model == 'product.product':
+            product = image_to_resequence
+            product_template = product.product_tmpl_id
+        elif image_res_model == 'product.template':
+            product_template = image_to_resequence
+            product = product_template.product_variant_id
+        else:
+            product = image_to_resequence.product_variant_id
+            product_template = product.product_tmpl_id or image_to_resequence.product_tmpl_id
+
+        if not product and not product_template:
+            raise ValidationError(_("Product not found"))
+
+        product_images = (product or product_template)._get_images()
+        if image_to_resequence not in product_images:
+            raise ValidationError(_("Invalid image"))
+
+        image_idx = product_images.index(image_to_resequence)
+        new_image_idx = 0
+        if move == 'left':
+            new_image_idx = max(0, image_idx - 1)
+        elif move == 'right':
+            new_image_idx = min(len(product_images) - 1, image_idx + 1)
+        elif move == 'last':
+            new_image_idx = len(product_images) - 1
+
+        # no-op resequences
+        if new_image_idx == image_idx:
+            return
+        # We can not move an embedded image to the first position (main product image)
+        if image_res_model == 'product.image' and image_to_resequence.embed_code and product_images[new_image_idx]._name != 'product.image':
+            raise ValidationError(_("Can not resequence embedded image/video with a non compatible image."))
+
+        # Swap images
+        other_image = product_images[new_image_idx]
+        source_field = hasattr(image_to_resequence, 'embed_code') and image_to_resequence.embed_code and 'embed_code' or 'image_1920'
+        target_field = hasattr(other_image, 'embed_code') and other_image.embed_code and 'embed_code' or 'image_1920'
+        other_image[target_field], image_to_resequence[source_field] = image_to_resequence[source_field], other_image[target_field]
```

Comment:
We also need to remove the possible `embed_code` from the source/target image in case it was not already an embedded image